### PR TITLE
feat(python): Adding CDK example chaining S3, SNS, SQS and Lambda

### LINKED ISSUE #520: CDK Example: S3 event -> SNS publish -> SQS subscription -> Lambda listener
## Overview
A common pattern that I have used to create a resilient way to trigger a lambda when an event occurs on S3 is to chain the following services:
- S3
- SNS
- SQS
- Lambda

## Justification
The reason I believe this is a good example is that it illustrates the following concepts:
- Chaining Services
- Disconnected Event Notifications and Handling
- Resilient storage to process request
- Asynchronous processing of request
- Scalable, Serverless Architecture
 
## Example
1. A user uploads a specific file type to an S3 Bucket
2. This upload triggers an event that notifies an SNS Topic by publishing a message with the S3 event details
3. The SNS Topic has an SQS queue that subscribes to notifications
4. When the upload event is published, the SNS message is put into the SQS Queue
5. A Lambda is polling the SQS as its Event Source
6. The Lambda receives the message from the SQS queue and processes it
7. If the message is not processed by the Lambda the message eventually ends up in a dead-letter queue

### Proposed Solution

I am proposing to build a CDK sample in Python that illustrates how to build and wire all these components together. I have the code for this solution written and tested and will be forking the repository to PR my solution.

### Language

Python

### Other information

_No response_

### Acknowledge

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by aabragan**
Fixes #520 

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

### CODE REVIEW

**Comment by aabragan** [comment]
@ryparker My sample was building fine. I just updated it so the merge could be smooth and now I am getting some weird CDK issue for the CDK 2.0 RC changes. This has nothing to do with my PR. Can we please get this PR merged as it has been sitting out here for a while.

```
aws-cdk-lib 2.0.0rc33 has requirement constructs<11.0.0,>=10.0.0, but you'll have constructs 3.3.161 which is incompatible.
25hInstalling collected packages: aws-cdk.cloud-assembly-schema, aws-cdk.cx-api, constructs, aws-cdk.region-info, aws-cdk.core, aws-cdk.assertions, aws-cdk.aws-iam, aws-cdk.aws-events, aws-cdk.assets, aws-cdk.aws-kms, aws-cdk.aws-s3, aws-cdk.aws-s3-assets, aws-cdk.aws-cloudwatch, aws-cdk.aws-logs, aws-cdk.aws-ssm, aws-cdk.aws-ec2, aws-cdk.aws-codestarnotifications, aws-cdk.aws-sqs, aws-cdk.aws-sns, aws-cdk.aws-ecr, aws-cdk.aws-codeguruprofiler, aws-cdk.aws-autoscaling-common, aws-cdk.aws-applicationautoscaling, aws-cdk.aws-signer, aws-cdk.aws-ecr-assets, aws-cdk.aws-efs, aws-cdk.aws-lambda, aws-cdk.aws-sns-subscriptions, aws-cdk.aws-cloudformation, aws-cdk.custom-resources, aws-cdk.aws-route53, aws-cdk.aws-certificatemanager, aws-cdk.aws-elasticloadbalancingv2, aws-cdk.aws-cognito, aws-cdk.aws-apigateway, aws-cdk.aws-kinesis, aws-cdk.aws-dynamodb, aws-cdk.aws-s3-notifications, aws-cdk.aws-sam, aws-cdk.aws-secretsmanager, aws-cdk.aws-lambda-event-sources, s3-sns-sqs-lambda-chain
  Found existing installation: constructs 10.0.9
    Uninstalling constructs-10.0.9:
      Successfully uninstalled constructs-10.0.9
  Running setup.py develop for s3-sns-sqs-lambda-chain
Successfully installed aws-cdk.assertions-1.126.0 aws-cdk.assets-1.126.0 aws-cdk.aws-apigateway-1.126.0 aws-cdk.aws-applicationautoscaling-1.126.0 aws-cdk.aws-autoscaling-common-1.126.0 aws-cdk.aws-certificatemanager-1.126.0 aws-cdk.aws-cloudformation-1.126.0 aws-cdk.aws-cloudwatch-1.126.0 aws-cdk.aws-codeguruprofiler-1.126.0 aws-cdk.aws-codestarnotifications-1.126.0 aws-cdk.aws-cognito-1.126.0 aws-cdk.aws-dynamodb-1.126.0 aws-cdk.aws-ec2-1.126.0 aws-cdk.aws-ecr-1.126.0 aws-cdk.aws-ecr-assets-1.126.0 aws-cdk.aws-efs-1.126.0 aws-cdk.aws-elasticloadbalancingv2-1.126.0 aws-cdk.aws-events-1.126.0 aws-cdk.aws-iam-1.126.0 aws-cdk.aws-kinesis-1.126.0 aws-cdk.aws-kms-1.126.0 aws-cdk.aws-lambda-1.126.0 aws-cdk.aws-lambda-event-sources-1.126.0 aws-cdk.aws-logs-1.126.0 aws-cdk.aws-route53-1.126.0 aws-cdk.aws-s3-1.126.0 aws-cdk.aws-s3-assets-1.126.0 aws-cdk.aws-s3-notifications-1.126.0 aws-cdk.aws-sam-1.126.0 aws-cdk.aws-secretsmanager-1.126.0 aws-cdk.aws-signer-1.126.0 aws-cdk.aws-sns-1.126.0 aws-cdk.aws-sns-subscriptions-1.126.0 aws-cdk.aws-sqs-1.126.0 aws-cdk.aws-ssm-1.126.0 aws-cdk.cloud-assembly-schema-1.126.0 aws-cdk.core-1.126.0 aws-cdk.custom-resources-1.126.0 aws-cdk.cx-api-1.126.0 aws-cdk.region-info-1.126.0 constructs-3.3.161 s3-sns-sqs-lambda-chain
+ /home/runner/work/aws-cdk-examples/aws-cdk-examples/scripts/synth.sh
+++ dirname /home/runner/work/aws-cdk-examples/aws-cdk-examples/scripts/synth.sh
++ cd /home/runner/work/aws-cdk-examples/aws-cdk-examples/scripts
++ pwd
+ scriptdir=/home/runner/work/aws-cdk-examples/aws-cdk-examples/scripts
+ export AWS_DEFAULT_REGION=us-east-1
+ AWS_DEFAULT_REGION=us-east-1
+ fake_context=false
+ '[' '!' -f cdk.context.json ']'
+ fake_context=true
+ true
+ cp /home/runner/work/aws-cdk-examples/aws-cdk-examples/scripts/fake.context.json cdk.context.json
+ npx cdk synth
Traceback (most recent call last):
  File "app.py", line 4, in <module>
    from aws_cdk import core as cdk
  File "/tmp/.venv/lib/python3.7/site-packages/aws_cdk/__init__.py", line 23759, in <module>
    from . import assertions
  File "/tmp/.venv/lib/python3.7/site-packages/aws_cdk/assertions/__init__.py", line 453, in <module>
    from ._jsii import *
  File "/tmp/.venv/lib/python3.7/site-packages/aws_cdk/assertions/_jsii/__init__.py", line 12, in <module>
    import aws_cdk.core._jsii
  File "/tmp/.venv/lib/python3.7/site-packages/aws_cdk/core/__init__.py", line 5412, in <module>
    class ConstructNode(metaclass=jsii.JSIIMeta, jsii_type="@aws-cdk/core.ConstructNode"):
  File "/tmp/.venv/lib/python3.7/site-packages/aws_cdk/core/__init__.py", line 5450, in ConstructNode
    runtime_info: typing.Optional[aws_cdk.cx_api.RuntimeInfo] = None,
AttributeError: module 'aws_cdk' has no attribute 'cx_api'
Subprocess exited with error 1
Error: Process completed with exit code 1.
```

**Comment by aabragan** [comment]
@njlynch see my comment above. It looks like the build is now failing after the CDK upgrades. This PR was ready for merge prior to that.

**Comment by aabragan** [comment]
@ryparker I just updated this sample to CDK v2 and all checks are passing. Can someone please review?

File: python/s3-sns-sqs-lambda-chain/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+aws-cdk-lib==2.12.0
+constructs>=10.0.0,<11.0.0
```
Comment (ryparker) [inline python/s3-sns-sqs-lambda-chain/requirements.txt:2]:
```suggestion
aws-cdk-lib>=2.12.0
constructs>=10.0.0
```
Lets keep this open to the latest CDK. This way we can use CI to identify when if it fails with a CDK update.

File: python/s3-sns-sqs-lambda-chain/README.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+> **This is a stable example. It should successfully build out of the box**
+>
+> This example is built on Construct Libraries marked "Stable" and does not have any infrastructure prerequisites to build.
+---
+<!--END STABILITY BANNER-->
+
+
+## Overview
+A common pattern that I have used to create a resilient way to trigger a lambda when an event occurs on S3 is to chain the following services:
```
Comment (ryparker) [inline python/s3-sns-sqs-lambda-chain/README.md:17]:
```suggestion
Create a resilient way to trigger a lambda when an event occurs on S3. It chains the following services:
```

File: python/s3-sns-sqs-lambda-chain/app.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env python3
+import os
+
+import aws_cdk as cdk
+
+from s3_sns_sqs_lambda_chain.s3_sns_sqs_lambda_chain_stack import S3SnsSqsLambdaChainStack
+
+
+app = cdk.App()
```
Comment (ryparker) [inline python/s3-sns-sqs-lambda-chain/app.py:9]:
```suggestion
import aws_cdk as cdk
from s3_sns_sqs_lambda_chain.s3_sns_sqs_lambda_chain_stack import S3SnsSqsLambdaChainStack

app = cdk.App()
```

Removed `os` since its not used. Removed some unnecessary spacing.

File: python/s3-sns-sqs-lambda-chain/cdk.json [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    "exclude": [
+      "README.md",
+      "cdk*.json",
+      "requirements*.txt",
+      "source.bat",
+      "**/__init__.py",
+      "python/__pycache__",
+      "tests"
+    ]
+  },
```
Comment (ryparker) [inline python/s3-sns-sqs-lambda-chain/cdk.json:16]:
```suggestion
```
For consistency with the other examples let's try to keep this config light by only including the required configurations.

File: python/s3-sns-sqs-lambda-chain/cdk.json [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    "@aws-cdk/aws-rds:lowercaseDbIdentifier": true,
+    "@aws-cdk/aws-lambda:recognizeVersionProps": true,
+    "@aws-cdk/aws-cloudfront:defaultSecurityPolicyTLSv1.2_2021": true,
+    "@aws-cdk-containers/ecs-service-extensions:enableDefaultLogDriver": true,
+    "@aws-cdk/aws-ec2:uniqueImdsv2TemplateName": true,
+    "@aws-cdk/core:target-partitions": [
+      "aws",
+      "aws-cn"
+    ]
+  }
```
Comment (ryparker) [inline python/s3-sns-sqs-lambda-chain/cdk.json:29]:
```suggestion
  "context": {
    "@aws-cdk/aws-apigateway:usagePlanKeyOrderInsensitiveId": true,
    "@aws-cdk/core:stackRelativeExports": true,
    "@aws-cdk/aws-rds:lowercaseDbIdentifier": true,
    "@aws-cdk/aws-cloudfront:defaultSecurityPolicyTLSv1.2_2021": true,
  }
```

Removed deprecated feature flags ([v2 feature flag docs](https://docs.aws.amazon.com/cdk/v2/guide/featureflags.html)). Are we sure that these remaining feature flags are necessary for this example? It would be nice if we could do without them.

File: python/s3-sns-sqs-lambda-chain/source.bat [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+rem
+rem     source .venv/bin/activate
+rem
+rem (which activates a Python virtualenv on Linux or Mac OS X) work on Windows.
+rem On Windows, this command just runs this batch file (the argument is ignored).
+rem
+rem Now we don't need to document a Windows command for activating a virtualenv.
+
+echo Executing .venv\Scripts\activate.bat for you
+.venv\Scripts\activate.bat
```
Comment (ryparker) [inline python/s3-sns-sqs-lambda-chain/source.bat:13]: Is this file necessary? We usually have some documentation in the README such as:

> If you are a Windows platform, you would activate the virtualenv like this:
>
> ```
> % .env\Scripts\activate.bat
> ```

File: python/s3-sns-sqs-lambda-chain/tests/unit/test_stack_s3_sns_sqs_lambda.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+
+def test_lambdas_created(template):
+    """
+    Test for Lambdas created:
+        - Sample Lambda
+        - Bucket Notification Handler (automatically provisioned)
+    """
+    template.resource_count_is("AWS::Lambda::Function", 2)
+    template.resource_count_is("AWS::Lambda::EventSourceMapping", 1)
```
Comment (ryparker) [inline python/s3-sns-sqs-lambda-chain/tests/unit/test_stack_s3_sns_sqs_lambda.py:67]: This is awesome. Thank you for adding tests.

File: python/s3-sns-sqs-lambda-chain/README.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+With specific profile,
+```
+$ cdk bootstrap --profile test
+```
+
+With the bootstrap complete, we are ready to deploy the lambda function and lambda layer.
+
+```
+$ cdk deploy
+```
```
Comment (ryparker) [inline python/s3-sns-sqs-lambda-chain/README.md:107]: Would you mind adding some documentation on how to run the tests? 🙏🏻

**Review by ryparker — CHANGES_REQUESTED** [review]
A few suggestions before we merge this.

Thank you for contributing with tests 🙏🏻

File: python/s3-sns-sqs-lambda-chain/s3_sns_sqs_lambda_chain/s3_sns_sqs_lambda_chain_stack.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+        )
+
+        function = _lambda.Function(self, "lambda_function",
+                                    runtime=_lambda.Runtime.PYTHON_3_9,
+                                    handler="lambda_function.handler",
+                                    code=_lambda.Code.from_asset(path=lambda_dir))
+
+        # This binds the lambda to the SQS Queue
+        invoke_event_source = lambda_events.SqsEventSource(upload_queue)
+        function.add_event_source(invoke_event_source)
```
Comment (ryparker) [inline python/s3-sns-sqs-lambda-chain/s3_sns_sqs_lambda_chain/s3_sns_sqs_lambda_chain_stack.py:94]:
```suggestion
        function.add_event_source(invoke_event_source)
        
         CfnOutput(self, "UploadFileToS3", value=f"aws s3 cp <local-path-to-file> s3: //${s3_bucket.bucketName} /", description="Upload a file to S3 (using AWS CLI) to trigger the SQS chain",
                  )
        CfnOutput(self, "UploadSqsQueue", value=upload_queue.queue_url, description="Link to Upload SQS Queue",
                  )
        CfnOutput(self, "LambdaFunction", value=function.functionName,
                  )
        CfnOutput(self, "LambdaFunctionLogs", value=function.logGroup.logGroupName,
                  )
```

Mind if we add some outputs? You'll have to add CfnOutput to the imports at the top BTW.

**Review by ryparker — CHANGES_REQUESTED** [review]

File: python/s3-sns-sqs-lambda-chain/s3_sns_sqs_lambda_chain/s3_sns_sqs_lambda_chain_stack.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
+from aws_cdk import (
+  Stack,
+  Duration,
+  aws_lambda as _lambda,
```
[the remark this answers is not in the document]
Comment (aabragan) [reply python/s3-sns-sqs-lambda-chain/s3_sns_sqs_lambda_chain/s3_sns_sqs_lambda_chain_stack.py:4]: @ryparker - `lambda` is a keyword in python and so this change will not work. I am not going to merge this change.

File: python/s3-sns-sqs-lambda-chain/source.bat (hunk 6 above)
Comment (aabragan) [reply python/s3-sns-sqs-lambda-chain/source.bat:13]: @ryparker this file is generated in the `cdk init` and so I would assume it was added by the team that built that template.

File: python/s3-sns-sqs-lambda-chain/README.md (hunk 8 above)
Comment (aabragan) [reply python/s3-sns-sqs-lambda-chain/README.md:107]: Sure thing, I will add that shortly.

File: python/s3-sns-sqs-lambda-chain/s3_sns_sqs_lambda_chain/s3_sns_sqs_lambda_chain_stack.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ryparker) [reply python/s3-sns-sqs-lambda-chain/s3_sns_sqs_lambda_chain/s3_sns_sqs_lambda_chain_stack.py:5]: Ah totally forgot about this 🤦🏻 . Never mind!

**Comment by aabragan** [comment]
@ryparker all comments for this PR have been addressed. Thanks!

File: python/s3-sns-sqs-lambda-chain/source.bat (hunk 6 above)
Comment (ryparker) [reply python/s3-sns-sqs-lambda-chain/source.bat:13]: Since we don't usually put that in our other examples and we have the same command referenced in the README would you mind deleting this file?

**Review by ryparker — APPROVED** [review]
Awesome. Thank you 🎉

**Review by ryparker — APPROVED** [review]

File: python/s3-sns-sqs-lambda-chain/source.bat (hunk 6 above)
Comment (ryparker) [reply python/s3-sns-sqs-lambda-chain/source.bat]: Got it
